---
title: Simulation exercise in<br />**Genomik und Bioinformatik**
output: html_document
---

```{r simulation_options, include=FALSE}
library(knitr)
opts_chunk$set(fig.width=5, fig.height=5, fig.path='RmdFigs/', dpi=72,
               warning=FALSE, message=FALSE, prompt=TRUE, cache=FALSE)
set.seed(53079239)
```

Part D) Programming - Normalization for microarray gene expression analysis\
**1. Load and normalize raw data:**

**Question (a):** Use the ReadAffy function from the affy package in R to read all CEL-files into a single R-object. How many samples does your object hold? How many features are reported? What does one feature actually stand for and on which scale are features measured? Use the method exprs to extract gene expression matrices from objects of type AffyBatch.

```{r Da}
library(affy)

myFiles <- list.files(path="C:/Users/aleX/Kurse/genomik/Blatt6/mmml/", pattern=".CEL.gz", full.names=TRUE )
length(myFiles)

myData <- ReadAffy(filenames=myFiles)
sampleNames(myData) <- sub(pattern="a?.CEL.gz", "", sampleNames(myData))

number_of_genes <- length(unique(row.names(myData)))
number_of_genes

myExpr <- exprs(myData)
summary(as.vector(myExpr))
```
**Answer:** Raw data contains more features than the rma-normalised data. This extra data is combined/summarised low level probe intensity values. By only counting the unique rownames or by later calling `str()` on the rma-normalised data set, we can see that there are 22283 genes and 176 samples. The scale can be viewed by `summary()` of the vectorized data set.

**Question (b):** Read the csv-file _MMML_pheno.csv_ from the same data directory. What does the matrix report? Integrate the matrix into the object returned by `ReadAffy` using the method `pData`. Why are the CEL-files and the csv-file compatible? Use the method `sampleNames` to collect sample names from an object of type `AffyBatch`.

```{r Db}
myCSV <- read.csv("C:/Users/aleX/Kurse/genomik/Blatt6/mmml/MMML_pheno.csv")
rownames(myCSV) <- myCSV[,1]
pData(myData) <- myCSV
#why does pheno fit into myData? because number of rows of phenoData is equal to number of columns of exprs
myphenoData <- pData(myData)
sampleNames(myData)[1:5]
```

**Answer:** The matrix reports the pheno data of the samples (i.e. gender, age, gcbabc). 
They are compatible because number of rows of `phenoData` is equal to number of columns of `exprs`.

**Question (c):** Normalize the raw Data twice, once with the function `rma` and once with the function `vsnrma`. Make sure to include `library(vsn)`. Compare the scales of the raw, the rma-normalized and the vsnrma-normalized data. You can extract the gene expression from the normalized objects using the `exprs` function.

```{r Dc}
library(vsn)

myExpr_rma <- exprs(rma(myData))
summary(as.vector(myExpr_rma))

myExpr_vsnrma <- exprs(vsnrma(myData))
summary(as.vector(myExpr_vsnrma))
```

**Answer:** The scales of the two normalization methods are roughly the same.

**2. Compare fold-changes from vsnrna-normalized data to fold-changes from rma-normalized data::**

**Question (a):** Consider samples of type GCB (germinal center B-cell like) and samples of type ABC (activated B-cell like) in the normalized lymphoma data sets to compute fold-changes between these classes. Where can you and the corresponding class labels per sample? How many GCB samples and how many ABC samples are available?

```{r Dd}
abc_index <- which(myphenoData$GCBABC == "ABC")
length(abc_index)

gcb_index <- which(myphenoData$GCBABC == "GCB")
length(gcb_index)
```
**Answer:** The class labels for _ABC_ and _GCB_ can be found in the pheno data.

**Question (b):** For each gene, compute a log fold-change2 of GCB versus ABC samples. Do this once for the data set normalized with `vsn` and once for the `rma` data set. Is taking the logarithm needed here to compute log fold-changes? Explain your decision. Is this different for the two normalization methods? What are the ranges of the fold-changes?

```{r De}
#rma
log2_avg_abc_rma <- apply(myExpr_rma[,abc_index], 1, mean)
log2_avg_gcb_rma <- apply(myExpr_rma[,gcb_index], 1, mean)
log2_fold_changes_rma <- log2_avg_abc_rma - log2_avg_gcb_rma
summary(log2_fold_changes_rma) #gives min, max fold-changes

#vsnrma
log2_avg_abc_vsnrma <- apply(myExpr_vsnrma[,abc_index], 1, mean)
log2_avg_gcb_vsnrma <- apply(myExpr_vsnrma[,gcb_index], 1, mean)
log2_fold_changes_vsnrma <- log2_avg_abc_vsnrma - log2_avg_gcb_vsnrma
summary(log2_fold_changes_vsnrma)
```

**Answer:** `Log()` of `rma` is not needed. From the documentation explains that: "Note that this expression measure is given to you in log base 2 scale. 
This differs from most of the other expression measure methods." Same is true for `vsnrma`.

**Question (f):** Plot the differences between gene-wise log fold-changes based on the two normalization methods. In addition, how strong are the differences between the fold-changes per gene? Find the 20 genes with the most different fold-changes compared between the two normalization methods. Are these genes among those with strong fold-changes?

```{r Df0, out.width = "100%"}
#plot diff between norm methods
plot(log2_fold_changes_rma, log2_fold_changes_vsnrma, pch=20, cex=0.5, ylab="rma", xlab="vsnrma", main="Difference between normalization methods")

#hist of log-fold diff of genes
changes_of_log2_fold_changes <- log2_fold_changes_rma - log2_fold_changes_vsnrma
hist(changes_of_log2_fold_changes, breaks=100, xlab="Difference of log2-fold changes of normalizations")
```

```{r Df1, echo = T, results = 'hide'}
#top 20 genes with rma fold-change bigger than vsnrma
most_diff_norm <- sort(changes_of_log2_fold_changes, decreasing=TRUE, index.return=TRUE)
most_diff_norm$x[1:20]
most_diff_norm$ix[1:20]

#top 20 genes with highest rma fold-change
most_diff_rma <- sort(log2_fold_changes_rma, decreasing=TRUE, index.return=TRUE) #order() would also work (instead of index.return=TRUE)
most_diff_rma$x[1:20]
most_diff_rma$ix[1:20]

count_rma=0
for(i in most_diff_norm$ix[1:20]){
  for(j in most_diff_rma$ix[1:20]){
    if(i == j){
      count_rma <- count_rma +1
    }
  }  
}
```
```{r Df11}
print(count_rma)
```
**Answer:** The goal is to find the 20 genes where the difference between the fold-change of the two normalization methods is the biggest. I did it once for the case that rma is bigger than vsnrma and vice versa (a bit below). We then find the top 20 genes with the highest fold-change in rma and vsnrma respectively. The question is now if those 20 genes do really have a high fold change or if it is artificially increased by the normalization method used. Therefore I compared the highest fold-change difference where rma is bigger than vsnrma with the top 20 genes with the highest fold-change of rma. If there is an overlap the fold-change is artifical. The result for rma was that in the top 20 genes there are 8 genes that "benefit" from the rma normalization.
The same was done for the vsnrma, where 0 out of the top 20 genes (with vsnrma fold-change normalization bigger than rma normalization) where found. So no overlap at all.

```{r Df2, echo = T, results = 'hide'}
#top 20 genes with vsnrma fold-change bigger than rma
most_diff_norm2 <- sort(changes_of_log2_fold_changes, decreasing=FALSE, index.return=TRUE)
most_diff_norm2$x[1:20]
most_diff_norm2$ix[1:20]

#top 20 genes with highest vsnrma fold-change
most_diff_vsnrma <- sort(log2_fold_changes_vsnrma, decreasing=TRUE, index.return=TRUE)
most_diff_vsnrma$x[1:20]
most_diff_vsnrma$ix[1:20]

count_vsnrma=0
for(i in most_diff_norm2$ix[1:20]){
  for(j in most_diff_vsnrma$ix[1:20]){
    if(i == j){
      count_vsnrma <- count_vsnrma +1
    }
  }  
}
```

```{r Df22}
print(count_vsnrma)
```

**Question (g):** Usually, researchers are most interested in genes with highest fold-changes. Thus, a list of genes sorted by fold-changes is of particular interest for biologists. The function rank takes a numerical vector and returns the rank for each element of the vector in a corresponding sorted list. Hence, you can easily compute positions for all genes in both lists of fold-changes for the different normalization methods. Illustrate the
positions per gene in these lists in a similar plot as used in question 2c and explain your observation.

```{r Dg, out.width = "100%"}
rank_rma <- rank(log2_fold_changes_rma)
rank_vsnrma <- rank(log2_fold_changes_vsnrma)

plot(rank_rma, rank_vsnrma, pch=20, cex=0.5, xlab="rma", ylab="vsnrma", main="difference between rank of normalizations")
lines(0:25000, 0:25000, col="red")
```

**Answer:** The ranks for both normalizations methods agree with each other. High ranks in rma are also high ranks in vsnrma. At the middle the two normalizations differ but this is to be expected since two different methods were used for normalization, namely multi-aray averaging (rma) and variance stabilization (vsn).

Information about the used R and package versions
-------------------------------------------------

The used versions of R and packages to generate the report above are
the following:

```{r sessionInfo}
sessionInfo()
```

